src/proposal.mligo

```diff
@@ -88,6 +105,44 @@ let update_delegates (params, store : update_delegate_params * storage): return
   , { store with delegates = List.fold update_delegate params store.delegates }
   )
 
+// TODO: Do we allow a call when voter does not exist or the actual votes is empty to go through?
+// TODO: Do we allow `unfreeze_staked` on multiple proposals at once?
+// TODO: Do we remove the old votes ?
+let unfreeze_staked (param, config, store : unfreeze_staked_param * config * storage): return =
+  let proposal_key = param in
+  let proposal = fetch_proposal (proposal_key, store) in
+
+  // Ensure proposal is already flushed or dropped.
+  let _ =  if proposal.is_proposer_unfreezed = true
+      then proposal
+      else (failwith("PROPOSAL_IS_NOT_FLUSHED") : proposal) in
+
+  // Check if voter exist.
+  let voter_proposal_map = match Map.find_opt Tezos.sender store.voters with
+        | Some m -> m
+        | None -> (failwith("VOTER_DOES_NOT_EXIST") : (proposal_key, vote_status list) map) in
+
+  // Get all that the votes on the proposal
+  let votes = match Map.find_opt proposal_key voter_proposal_map with
+        | Some v -> v
+        | None -> (failwith("NO_VOTE_ON_PROPOSAL") : vote_status list) in
+
+  // Do the unstake
+  let combine_tokens = fun (acc, vote : nat * vote_status) ->
+    acc + vote.vote_amount
+  in
+  let tokens = List.fold combine_tokens votes 0n in
+  let store = unstake_tk(tokens, 0n, Tezos.sender, config.period, store) in
+
+  // Remove the votes
+  let new_voter_proposal_map = Map.update proposal_key (None : (vote_status list) option) voter_proposal_map in
+
+  ( nil_op
+  , { store with
+      voters = Big_map.update Tezos.sender (Some new_voter_proposal_map) store.voters
+    }
+  )
+
```

Comment:
@pasqu4le This PR is almost ready for review, only the tests are left to be done.

Before updating the tests, I want your advice on the current implementation. The new entrypoint is call `Unfreeze_unstaked`. Currently it only targets one proposal at a time, my questions are:
1. Do we plan to allow the user to unstake all their votes (on flushed proposals) all at once or not?
2. If the voter that call does not actually vote on that proposal, do we want to fail the call or let it through (by unfreeze nothing)? 
3. Do we keep their votes as history after `Unfreeze_unstaked` is called or removed them?

Hope to get your advice on this.

Reply:
@pasqu4le I agree with all your points. I have push the final changes as well as updating the tests now, this should be ready for review. The only thing left is updating the docs, which will be what I'm working on next.

The new entrypoint is call `Unfreeze_staked`, let me know if the name is good enough as well.

Reply:
@rinn7e what do you think of `unfreeze_vote` as a name instead, given that it can only be used for tokens that were frozen for voting?